tempesta_fw/sock.c

```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
+					goto unlock;
+				}
+				dev = dev_get_by_index_rcu(net, sk->sk_bound_dev_if);
+				if (!dev) {
+					err = -ENODEV;
+					goto unlock;
+				}
+			}
+
+			/* ipv4 addr of the socket is invalid.  Only the
+			 * unspecified and mapped address have a v4 equivalent.
+			 */
+			v4addr = LOOPBACK4_IPV6;
+			if (!(addr_type & IPV6_ADDR_MULTICAST))	{
+				if (!(inet->freebind || inet->transparent) &&
+				    !ipv6_chk_addr(net, &addr->sin6_addr,
+						   dev, 0)) {
+					err = -EADDRNOTAVAIL;
+					goto unlock;
+				}
+			}
+			rcu_read_unlock();
+		}
+	}
+
+	inet->inet_rcv_saddr = v4addr;
+	inet->inet_saddr = v4addr;
+
+	np->rcv_saddr = addr->sin6_addr;
+
+	if (!(addr_type & IPV6_ADDR_MULTICAST))
+		np->saddr = addr->sin6_addr;
+
+	/* Make sure we are allowed to bind here. */
+	if (sk->sk_prot->get_port(sk, snum)) {
+		inet_reset_saddr(sk);
+		err = -EADDRINUSE;
+		goto out;
+	}
+
+	if (addr_type != IPV6_ADDR_ANY) {
+		sk->sk_userlocks |= SOCK_BINDADDR_LOCK;
+		if (addr_type != IPV6_ADDR_MAPPED)
+			np->ipv6only = 1;
+	}
+	if (snum)
+		sk->sk_userlocks |= SOCK_BINDPORT_LOCK;
+	inet->inet_sport = htons(inet->inet_num);
+	inet->inet_dport = 0;
+	inet->inet_daddr = 0;
+unlock:
+	rcu_read_unlock();
+out:
+	return err;
+}
+
 int
 ss_bind(struct sock *sk, struct sockaddr *uaddr, int uaddr_len)
 {
-	struct socket sock = {
-		.sk = sk,
-		.type = sk->sk_type
-	};
+	int err;
+
 	BUG_ON((sk->sk_family != AF_INET) && (sk->sk_family != AF_INET6));
 	BUG_ON(sk->sk_type != SOCK_STREAM);
+
+	bh_lock_sock(sk);
 	if (sk->sk_family == AF_INET)
-		return inet_bind(&sock, uaddr, uaddr_len);
+		err = ss_inet_bind(sk, uaddr, uaddr_len);
 	else
-		return inet6_bind(&sock, uaddr, uaddr_len);
+		err = ss_inet6_bind(sk, uaddr, uaddr_len);
+
+	bh_unlock_sock(sk);
+	return err;
 }
 EXPORT_SYMBOL(ss_bind);
 
-/*
- * The original function is inet_listen() that is common to IPv4 and IPv6.
- * There isn't much to make shorter there, so just invoke it directly.
+ /*
+ * Move a socket into listening state. The original functions inet_listen().
+ *
+ * NOTE: This code assumes that both IPv4 and IPv6 are compiled in as
+ * part of the Linux kernel, and not as separate loadable kernel modules.
  */
+static int ss_inet_listen(struct sock *sk, int backlog)
+{
+	unsigned char old_state;
+	int err;
+
+	old_state = sk->sk_state;
+	if (!((1 << old_state) & (TCPF_CLOSE | TCPF_LISTEN)))
+		goto out;
+
+	/* Really, if the socket is already in listen state
+	 * we can only allow the backlog to be adjusted.
+	 */
+	if (old_state != TCP_LISTEN) {
+		err = inet_csk_listen_start(sk, backlog);
+		if (err)
+			goto out;
+	}
+	sk->sk_max_ack_backlog = backlog;
+	err = 0;
+
+out:
+	return err;
+}
+
 int
 ss_listen(struct sock *sk, int backlog)
 {
-	struct socket sock = {
-		.sk = sk,
-		.type = sk->sk_type,
-		.state = SS_UNCONNECTED
-	};
+	int err;
+
 	BUG_ON(sk->sk_type != SOCK_STREAM);
-	return inet_listen(&sock, backlog);
+
+	bh_lock_sock(sk);
+	err = ss_inet_listen(sk, backlog);
```

Comment:
`ss_inet_listen()` is never called in any context other than from `ss_listen()`. Therefore there's no reason to have it as a separate function